lib/active_merchant/billing/gateways/commerce_hub.rb

```diff
@@ -0,0 +1,327 @@
+module ActiveMerchant #:nodoc:
+  module Billing #:nodoc:
+    class CommerceHubGateway < Gateway
+      self.test_url = 'https://cert.api.fiservapps.com/ch'
+      self.live_url = 'https://prod.api.fiservapps.com/ch'
+
+      self.supported_countries = ['US']
+      self.default_currency = 'USD'
+      self.supported_cardtypes = %i[visa master american_express discover]
+
+      self.homepage_url = 'https://developer.fiserv.com/product/CommerceHub'
+      self.display_name = 'CommerceHub'
+
+      STANDARD_ERROR_CODE_MAPPING = {}
+
+      SCHEDULED_REASON_TYPES = %w(recurring installment)
+      ENDPOINTS = {
+        'sale' => '/payments/v1/charges',
+        'authonly' => '/payments/v1/charges',
+        'capture' => '/payments/v1/charges',
+        'void' => '/payments/v1/cancels',
+        'refund' => '/payments/v1/refunds',
+        'verify' => '/payments-vas/v1/accounts/verification',
+        'vault' => '/payments-vas/v1/tokens'
+      }
+
+      def initialize(options = {})
+        requires!(options, :api_key, :api_secret, :merchant_id, :terminal_id)
+        super
+      end
+
+      def purchase(money, payment, options = {})
+        post = {}
+        options[:capture_flag] = true
+        add_transaction_details(post, options)
+        build_purchase_and_auth_request(post, money, payment, options)
+
+        commit('sale', post, options)
+      end
+
+      def authorize(money, payment, options = {})
+        post = {}
+        options[:capture_flag] = false
+        add_transaction_details(post, options)
+        build_purchase_and_auth_request(post, money, payment, options)
+
+        commit('authonly', post, options)
+      end
+
+      def capture(money, authorization, options = {})
+        commit('capture', post, options)
+      end
+
+      def refund(money, authorization, options = {})
+        post = {}
+        add_invoice(post, money, options) if money
+        add_transaction_details(post, options)
+        add_reference_transaction_details(post, authorization, options)
+        commit('refund', post, options)
+      end
+
+      def void(authorization, options = {})
+        post = {}
+        add_transaction_details(post, options)
+        add_reference_transaction_details(post, authorization, options)
+        commit('void', post, options)
+      end
+
+      def store(credit_card, options = {})
+        post = {}
+        add_payment(post, credit_card, options)
+        add_billing_address(post, credit_card, options)
+        add_transaction_details(post, options)
+        add_transaction_interaction(post, options)
+        commit('vault', post, options)
+      end
+
+      def verify(credit_card, options = {})
+        MultiResponse.run(:use_first_response) do |r|
+          r.process { authorize(100, credit_card, options) }
+          r.process(:ignore_result) { void(r.authorization, options) }
+        end
+      end
```

Comment:
This gateway supports $0 auth with the `charges` endpoint. We should implement this instead of the traditional auth/void. Here's the [documentation](https://developer.fiserv.com/product/CommerceHub/docs/?path=docs/Resources/API-Documents/Payments/Charges.md&branch=main).

Reply:
Thanks @naashton, the change is updated 👍